R/data.table.R

```diff
@@ -1892,73 +1890,52 @@ as.matrix.data.table = function(x, rownames=NULL, rownames.value=NULL, ...) {
+             brackify(dim(rownames.value)))
+      
+      # Warn if list column:
+      if (is.list(rownames.value))
+        warning("x[,", rownames, "] is a list column, which will be coerced to a",
+                " character vector when used as rownames")
+    }
   } else if (!is.null(rownames.value)) {
     if (length(rownames.value)!=nrow(x))
       stop("length(rownames.value)==", length(rownames.value),
            " but should be nrow(x)==", nrow(x))
   }
+  
+  # Get matrix meta-data
+  cn = names(x)
   if (!is.null(rownames)) {
-    # extract that column and drop it.
-    rownames.value = x[[rownames]]
-    dm = dim(x) - 0:1
-    cn = names(x)[-rownames]
-    X = x[, .SD, .SDcols = cn]
-  } else {
-    dm = dim(x)
-    cn = names(x)
-    X = x
-  }
-  if (any(dm == 0L))
-    return(array(NA, dim = dm, dimnames = list(rownames.value, cn)))
-  p = dm[2L]
-  n = dm[1L]
-  collabs = as.list(cn)
-  class(X) = NULL
-  non.numeric = non.atomic = FALSE
-  all.logical = TRUE
-  for (j in seq_len(p)) {
-    if (is.ff(X[[j]])) X[[j]] = X[[j]][]   # nocov to bring the ff into memory, since we need to create a matrix in memory
-    xj = X[[j]]
-    if (length(dj <- dim(xj)) == 2L && dj[2L] > 1L) {
-      if (inherits(xj, "data.table"))
-        xj = X[[j]] = as.matrix(X[[j]])
-      dnj = dimnames(xj)[[2L]]
-      collabs[[j]] = paste(collabs[[j]], if (length(dnj) >
-        0L)
-        dnj
-      else seq_len(dj[2L]), sep = ".")
-    }
-    if (!is.logical(xj))
-      all.logical = FALSE
-    if (length(levels(xj)) > 0L || !(is.numeric(xj) || is.complex(xj) || is.logical(xj)) ||
-        (!is.null(cl <- attr(xj, "class", exact=TRUE)) && any(cl %chin%
-        c("Date", "POSIXct", "POSIXlt"))))
-      non.numeric = TRUE
-    if (!is.atomic(xj))
-      non.atomic = TRUE
-  }
-  if (non.atomic) {
-    for (j in seq_len(p)) {
-      xj = X[[j]]
-      if (is.recursive(xj)) { }
-      else X[[j]] = as.list(as.vector(xj))
-    }
-  }
-  else if (all.logical) { }
-  else if (non.numeric) {
-    for (j in seq_len(p)) {
-      if (is.character(X[[j]])) next
-      xj = X[[j]]
-      miss = is.na(xj)
-      xj = if (length(levels(xj))) as.vector(xj) else format(xj)
-      is.na(xj) = miss
-      X[[j]] = xj
-    }
-  }
-  X = unlist(X, recursive = FALSE, use.names = FALSE)
-  dim(X) <- c(n, length(X)/n)
-  dimnames(X) <- list(rownames.value, unlist(collabs, use.names = FALSE))
-  X
+    if(is.character(rownames)) cn = cn[-chmatch(rownames, cn)]
+    else cn = cn[-rownames]
+  }
+  p = length(cn)
+  n = nrow(x)
+  
+  # If no rows or columns can simply return empty array
+  if (p == 0L || n == 0L)
+    return(array(NA, dim = list(n, p), dimnames = list(rownames.value, cn)))
+  
+  # Put each column inside a list so we can use rbindlist to stack into a single vector
+  l = replicate(p, vector("list", 1), simplify=FALSE) 
+  for (j in seq_len(p))
+    l[[j]][[1]] = x[[j]] # copies pointer to column 
```

Comment:
Actually quite straightforward to get data.table into the format needed without needing to modify Crbindlist with a special argument.

This copies the pointer to each column into the appropriate nested list object, i.e. it is the equivalent of:

```
SEXP l = PROTECT(allocVector(VECSXP, p));
for (int j = 0; j < p; ++j) {
   SEXP thisCol = PROTECT(allocVector(VECSXP, 1)); 
   SET_VECTOR_ELT(thisCol, 0, VECTOR_ELT(x, j));
}
```